### PR TITLE
Error on redeclarations of undefined

### LINKED ISSUE #3087: Type alias named `undefined` is allowed, but doesn't work
```ts
type undefined = string;
var x: undefined; // Error, cannot find name 'undefined'
```
Expected: Type alias should be an error, or the type annotation should work

### PR DESCRIPTION
**Pull request by weswigham**
From discussion on #5387.

Instead of this variety of implementation, another other option is actually exposing the `undefined` type as a primitive-like type (like `symbol` or `number`) and adding `declare var undefined: undefined` to the lib. That would be more consistent - and could begin to pave the way to more easily explicitly "nullable" types. 

Fixes #3087.

### CODE REVIEW

File: src/compiler/diagnosticMessages.json [hunk 1]
```diff
@@ -1168,6 +1168,10 @@
         "category": "Error",
         "code": 2396
     },
+    "Declaration duplicates builtin global identifier '{0}'.": {
```
Comment (yuit) [inline src/compiler/diagnosticMessages.json]: just curious, why not use `Duplicate identifier '{0}'` error?

Comment (DanielRosenwasser) [inline src/compiler/diagnosticMessages.json]: I'd prefer a more helpful message. This is the better direction IMO.

Though, `:%s/duplicates/name conflicts` `:%s/builtin/built-in`. P.S. I don't know if those are valid vim replacement commands.

Comment (yuit) [inline src/compiler/diagnosticMessages.json]: I am fine with the direction. Just curious and I do like the new error :)

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -336,6 +341,13 @@ namespace ts {
                         target[id] = source[id];
                     }
                     else {
+                        if (target === globals && source !== builtinGlobals) {
```
Comment (mhegazy) [inline src/compiler/checker.ts]: so why not just let them merge, then later on make sure that the undefined symbol is the one we have created, or that it does not have any declarations?

Comment (weswigham) [inline src/compiler/checker.ts]: We can get some wonky undefined behavior while typechecking if we actually let them merge (eg, this is a class which is also `undefined`), so we have to guard against it in the merge function so we don't let it poison the type checking process. Outright overriding the symbol is _okay_ and is what #5387 allowed, but merging with the builtin undefined symbol is unallowable.

Comment (mhegazy) [inline src/compiler/checker.ts]: This is a corner case, if the user ran into it, we just should let them know, i do not think we need to complicate our response here.

Comment (weswigham) [inline src/compiler/checker.ts]: By "wonky undefined behavior" I mean crashing.

**Comment by yuit** [comment]
:+1:

**Pull request merged**